### ISSUE TITLE
Feature Request: Jump directly to the implementation of a method in Java

### ISSUE THREAD

**Issue by ahnick** [ervandew/eclim#349]
So I've used :JavaSearchContext to jump directly to the declaration of a method; however, what I want to do is jump to the _implementation_.   

In Eclipse you can do this by holding down the Ctrl key(Command key on Mac) and hovering over a method name with your mouse and then you are given a menu option to open the following
- Open Declaration
- Open Implementation
- Open Return Type

When you select "Open Implementation" it will jump you directly to the file that implements that method.

Another way to do this in Eclipse is to open the declaration of the method and then hit "F4" where you will be able to see the Type Hierarchy for the current class you are in and then you can select the class that implements the method you are looking for.

I've looked all through the documentation for this, but I have not been able to locate a way to do this.

**Comment by cosminadrianpopescu** [comment]
Sorry, but it still doesn't. I am using eclim 2.4.1. It doesn't find any results.

**Comment by dhleong** [comment]
You're going to have to provide a minimal example that demonstrates that.
@ervandew's merged code works for everyone else's cases.

See discussion here:  https://github.com/ervandew/eclim/pull/358

**Comment by cosminadrianpopescu** [comment]
I am having a big project, with 8 eclipse projects (for dao, services, war etc.). If I just do `:JavaSearch -p MyClass.MyMethod -t method -x implementors`, then it will go to the desired method. However, If I just do `:JavaSearch -p MyMethod` this will return no result. Not even a list of possible results (`MyMethod` exists in several classes).

**Comment by dhleong** [comment]
The code and the feature are for the `-x implementors` flag, so it sounds
like there's nothing wrong related to this ticket.

I assume you meant `-p MyClass.MyMethod` in your counter example? Haven't
looked at that flag but I suspect without being qualified with the class
name it wouldn't find what you're looking for.

Have you tried simply placing the cursor over the method name and omitting
-p from the command? I never use that flag, myself. Also, if you don't want
implementations, you can try `-x declarations`, I believe.

**Comment by cosminadrianpopescu** [comment]
That's exactly the problem. I need to do `:JavaSearch MyClass.MyMethod -x implementors`. If I look to the `eclim#java#SearchAndDisplay` function I can see that the function `eclim#Execute` is called with a different command for `JavaSearch` and `JavaSearchContext`. For `JavaSearchContext` only a pattern is sent, and then it's normal that it requires the class name. But, take for example the following code: 

```
String s = myClass.MyMethod(params)
```

Now, if I am with the cursor on top of the `MyMethod` word, is kinda difficult to map something there to tell eclim that I am referring to MyClass class. 

While the other function, `JavaSearchContext` is sending a command indicating to eclim the position in the file, and the file in which we are executing the autocomplete. This way, eclim will know exactly from which class the method `MyMethod` comes. But the problem is that this function does not accept the `-x` flag. 

This is why I need that workaround. Once I do `JavaSearchContext`, vim will open the file containing the class in which `MyMethod` is defined. From there, if the file is called `MyClass.java`, I can assume that the class name is `MyClass` (so, the need for `fnamemodify` call) and then do the next call with the current word (`expand(<cword>)`) adding the name of the class. 

If the `JavaSearchContext` command would accept the `-x` flag and take vim directly to the implementation, this would be great. Otherwise, I will continue to use this workaround or I will keep searching in the source code, maybe I find a better work around.

**Comment by dhleong** [comment]
I'm not sure what is causing your difficulty, but I think there is some
misunderstanding. I'm not sure where you see that a different command is
used. `JavaSearchContext` is a convenience for `JavaSearch`—If you look at
ftplugin/java.vim, both pass the same args to
`eclim#java#search#SearchAndDisplay`, so they both use the `java_search`
command. The difference is that `JavaSearchContext` just tries to figure
out which `-x` flag you meant based on the element under the cursor. If you
omit the `-p` flag, both methods will use the element under the cursor.

It is not difficult to figure out what method is under the cursor; Eclipse
has a model of the source document and knows *exactly* what's under it. I
just tried doing `:JavaSearch` with no args on a static method as in your
example and it jumped to exactly the right place. Perhaps I'm still
misunderstanding your use case, but it seems like you're trying to find the
definition of a static method. For that case, `-x implementors` is not what
you want. That flag is for locating implementations of an *abstract*
method.

Perhaps you are using an old version of Eclim? I noticed you referenced
`eclim#java#SearchAndDisplay` instead of
`eclim#java#search#SearchAndDisplay`. Was that just a typo?

**Comment by cosminadrianpopescu** [comment]
That was just a typo. I am using eclim 2.4.1 which I installed a few days ago. `JavaSearchContext` is not a convenience for `JavaSearch`. Check the `eclim#java#search#SearchAndDisplay` function. This is calling the `s:Search` function defined in the `eclim/java/search.vim` file. 

This function has a big `if`: 

```
  if argline !~ '-p\>'
```

So, basically, if we are calling `JavaSearchContext` we will be on the main branch of this `if`. If we are calling `JavaSearch` we are going to be on the else branch of this If. So, doing `JavaSearch -p`, the command sent to eclim is something like this: 

```
-command java_search -n "spring-framework-2.5" -f "src/org/springframework/aop/aspectj/AbstractAspectJAdvice.java" -p "getUserAttribute"
``` 

You can notice that no position is sent regarding the file in which the search is done, so from here we can understand the need to search by prefixing with a class. 

But doing `JavaSearchContext` with the cursor on top of the `getUserAttribute` method, the command sent to the server is 

```
-command java_search -n "spring-framework-2.5" -f "src/org/springframework/aop/aspectj/AbstractAspectJAdvice.java" -o 24521 -e utf-8 -l 16 
``` 
You can see that now the position in the file is sent to `eclim`. So now, eclim will know that the method `getUserAttribute` is part of the `ProxyMethodInvocation` (examples from the spring framework source code). The problem is that `JavaSearchContext` will just ignore the `-x` flag and will always redirect vim to the interface. From there, using the workaround described above, I can reach the implementation of the method with another `vim` mapping. 

Basically I would like for the `JavaSearchContext` to take into consideration the `-x` flag.

**Comment by dhleong** [comment]
> JavaSearchContext is not a convenience for JavaSearch

Yes, it is:

```vim
if !exists(":JavaSearch")
  command -buffer -nargs=*
    \ -complete=customlist,eclim#java#search#CommandCompleteSearch
    \ JavaSearch :call eclim#java#search#SearchAndDisplay('java_search', '<args>')
endif
if !exists(":JavaSearchContext")
  command -buffer -nargs=?
    \ -complete=customlist,eclim#java#search#CommandCompleteSearchContext
    \ JavaSearchContext :call eclim#java#search#SearchAndDisplay('java_search', '<args>')
endif
```

The only difference is the completion function.

> So, doing JavaSearch -p, the command sent to eclim is something like this:

No position is sent *because you are using `-p`*. You don't need to; it's optional. Try *just* using `:JavaSearch` with no args and it will probably go where you want. If you are looking an abstract method, then `:JavaSearch -x implementors` will do what you expect.

**Comment by cosminadrianpopescu** [comment]
I agree with you that the command is calling the same function. But the `<args>` is different for the calls. Inside this function (`SearchAndDisplay`) there is a distinction depending on the two totally different sets of arguments. 

```
  " check if just a pattern was supplied.
  if argline =~ '^\s*\w'
    let argline = '-p ' . argline
  endif
```

So, doing `JavaSearch` even without `-p`, the `-p` will be added automatically. So `JavaSearch` is the same with or without `-p`

The position is sent when not using `-p`. Please check the two commands. The one from `JavaSearchContext` is sending `-o 24251`, which is the position inside the file. This is very important. Without the position we don't know from which class the method is coming.  

I am speaking about a method from a class which implements an interface. In this case, `JavaSearchContext` will take me to the interface, while `JavaSearch -p` needs to be prefixed with the class. This is not optimal, because prefixing with a class name cannot be done using a vim mapping. I am trying to go to the definition using a vim mapping. 

I can always do `JavaSearch -p MyClass.MyMethod -x implementors`. But this cannot be mapped to a key in `vim`, since is complicated to extract the name of the class and put it into a mapping. For this I have the workaround. 

But it would be nice to jump directly to the function definition, not the interface, by using a single `vim` command. The big difference between `JavaSearchContext` and `JavaSearch` is that `JavaSearchContext` is providing the position in the current file to eclim . See the `-o` parameter. This way, eclim will find by itself the class and will redirect. 

Thank you for your support and your patience.

**Comment by dhleong** [comment]
> So JavaSearch is the same with or without -p

No, it's not. You're misreading that. If you do `:JavaSearch hello` then it will interpret that as `:JavaSearch -p hello`. If you do `:JavaSearch` by itself, or with `-x implementors`, it will *not* do a pattern search. Let me prove it to you.

```java
Daniels-iMac:eclim dhleong$ git diff
diff --git a/org.eclim.jdt/java/org/eclim/plugin/jdt/command/search/SearchCommand.java b/org.eclim.jdt/java/org/eclim
index 21deb94..6c5e91a 100644
--- a/org.eclim.jdt/java/org/eclim/plugin/jdt/command/search/SearchCommand.java
+++ b/org.eclim.jdt/java/org/eclim/plugin/jdt/command/search/SearchCommand.java
@@ -38,6 +38,8 @@ import org.eclim.annotation.Command;
 import org.eclim.command.CommandLine;
 import org.eclim.command.Options;
 
+import org.eclim.logging.Logger;
+
 import org.eclim.plugin.core.command.AbstractCommand;
 
 import org.eclim.plugin.core.util.ProjectUtils;
@@ -88,6 +90,9 @@ import org.eclipse.jdt.core.search.SearchPattern;
 public class SearchCommand
   extends AbstractCommand
 {
+    private static final Logger logger =
+        Logger.getLogger(SearchCommand.class);
+
 
   public static final String CONTEXT_ALL = "all";
   public static final String CONTEXT_DECLARATIONS = "declarations";
@@ -166,6 +171,8 @@ public class SearchCommand
     String length = commandLine.getValue(Options.LENGTH_OPTION);
     String pat = commandLine.getValue(Options.PATTERN_OPTION);
 
+    logger.info("Offset: " + offset + "; pattern: " + pat);
+
     SearchPattern pattern = null;
     IJavaProject javaProject = project != null ?
       JavaUtils.getJavaProject(project) : null;
```

I add a log that dumps the "offset" and "pattern" arguments. First:

    :JavaSearch

> 2015-03-02 15:23:45,155 INFO  [org.eclim.plugin.jdt.command.search.SearchCommand] Offset: 3677; pattern: null

Next:

    :JavaSearch -x implementors

> 2015-03-02 15:25:13,516 INFO  [org.eclim.plugin.jdt.command.search.SearchCommand] Offset: 3677; pattern: null

Now:

    :JavaSearch hello

> 2015-03-02 15:25:45,362 INFO  [org.eclim.plugin.jdt.command.search.SearchCommand] Offset: null; pattern: hello

And of course:

    :JavaSearch -p hello

> 2015-03-02 15:26:07,240 INFO  [org.eclim.plugin.jdt.command.search.SearchCommand] Offset: null; pattern: hello


Do you believe me now? You *do not need `-p`*.

If you try `:JavaSearch -x implementors` on an abstract method reference (eg: `ISomething foo = /* .. */; foo.doBar()`, cursor on `doBar`) and you get no results, then that would be a bug with the `-x implementors` implementation that @ervandew did, and could perhaps be discussed in this ticket. If you do `:JavaSearch` on anything and it doesn't find declarations, that would also be a bug, but should be filed separately since it's unrelated to this ticket.

**Comment by cosminadrianpopescu** [comment]
I do believe you, of course. The problem is that this does not work. `JavaSearch -x implementors` with the cursor on top of my function does not return any result (`No results for 'getContract'`). The command sent to `eclim` is this: `-command java_search -n "immogest-war" -f "src/main/...java" -o 6089 -e utf-8 -l 11 -x implementors`. 

So you are right. But also, doing `JavaSearchContext -x implementors` returns the same command. The problem is that this return no results, while ``-command java_search -n "immogest-war" -f "src/main/...java" -o 6089 -e utf-8 -l 11` (exactly the same command but without -x implementors` takes me to the interface definition. 

So I suppose I should open a bug? 

You are telling me that `JavaSearchContext -x implementors` should work?

**Comment by dhleong** [comment]
I think what you want is just `:JavaSearch -x implementors`. Based on the
vim declaration of `:JavaSearchContext`, (using -nargs=?) it should only
accept at most the optional string pattern.

With that said, yes, `-x implementors` should work, if the `getContract()`
method is abstract and the implementation is on the classpath of the
project. I assume you've double-checked that Eclipse by itself will give
you the right thing if you ctrl-click and select "Open Implementations"? If
that works but Eclim doesn't, then @ervandew will probably need a minimal
project (or projects) that reproduces the bug. I would file it separately.

**Comment by cosminadrianpopescu** [comment]
Yes, of course eclipse works. Also vim eclim works as described above: I can do `JavaSearch -p MyClass.getContract -t method -x implementors` and this will work. The problem is that I need to prefix it with the class and by doing this, I cannot have a vim mapping to take me directly to the implementation. 

I go first to definition (`JavaSearchContext`), and from there I assume that the name of the file is also the name of the class and I can have one vim mapping to execute the `JavaSearch -p` command. 

Anyway, thank you very much for your support.

**Comment by alnavv** [comment]
This works, will take the method under the cursor and will look for the implemenation (not neccesary to prefix with tthe class name):

nnoremap <leader>jsi :JavaSearch -p <C-R><C-W> -t method -x implementors<CR>

**Comment by prathyk** [comment]
@alnavv , this only works if there exists one implementation of that method. If there is a different class hierarchy where the same method is implemented, this shows all possible implementations of other class hierarchies and not only the ones related to the current context

**Comment by jackiilong** [comment]
nnoremap <silent> <leader>m :JavaSearch -p <c-r>=expand("<cword>")<cr> -t method -x implementors<CR>"

this works.
